lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,203 @@
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * This class should only be used through {@link FailureCheckingLockWatchEventCache} as a proxy; failure to do so will
+ * result in concurrency issues and inconsistency in the cache state.
+ */
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final TimestampToVersionMap timestampMap;
+
+    public static LockWatchEventCache create() {
+        return FailureCheckingLockWatchEventCache.newProxyInstance(
+                new LockWatchEventCacheImpl(ClientLockWatchEventLogImpl.create()));
+    }
+
+    @VisibleForTesting
+    LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+        timestampMap = new TimestampToVersionMap();
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public void processStartTransactionsUpdate(
+            Set<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+        latestVersion.ifPresent(version -> startTimestamps.forEach(timestamp -> timestampMap.put(timestamp, version)));
+
+        getEarliestVersion().ifPresent(eventLog::removeOldEntries);
+    }
+
+    @Override
+    public void processGetCommitTimestampsUpdate(
+            Collection<TransactionUpdate> transactionUpdates,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+        latestVersion.ifPresent(version -> transactionUpdates.forEach(
+                transactionUpdate -> checkConditionOrThrow(!timestampMap.replace(transactionUpdate, version),
+                        "start timestamp missing from map")));
+    }
+
+    @Override
+    public CommitUpdate getCommitUpdate(long startTs) {
+        Optional<MapEntry> maybeEntry = timestampMap.get(startTs);
+        Optional<CommitInfo> maybeCommitInfo = maybeEntry.flatMap(MapEntry::commitInfo);
+
+        checkConditionOrThrow(!maybeCommitInfo.isPresent(), "commit info not processed for start timestamp");
+
+        CommitInfo commitInfo = maybeCommitInfo.get();
+
+        ClientLogEvents update =
+                eventLog.getEventsBetweenVersions(Optional.of(maybeEntry.get().version()), commitInfo.commitVersion());
+
+        if (update.clearCache()) {
+            return ImmutableInvalidateAll.builder().build();
+        }
+
+        return constructCommitUpdate(commitInfo, update.events());
+    }
+
+    @Override
+    public TransactionsLockWatchEvents getEventsForTransactions(
+            Set<Long> startTimestamps,
+            Optional<IdentifiedVersion> startVersion) {
+        Preconditions.checkArgument(!startTimestamps.isEmpty(), "Cannot get events for empty set of tranasctions");
```

Comment:
Just wondering if perhaps, for the sake of atlasdb-proxy, that we should fall through on this case instead (i.e. return no updates, and an empty map).

Reply:
Let's go with what you have and see how we handle this upstream. And adjust if we need to

Reply:
Ok, lets see how that goes.